lib/Sema/CSApply.cpp

```diff
@@ -107,6 +107,17 @@ Type Solution::computeSubstitutions(
   auto lookupConformanceFn =
       [&](CanType original, Type replacement, ProtocolType *protoType)
           -> ProtocolConformanceRef {
+    // Replacing type variable with archetype (if possible) is required
+    // here because type variables are never going to produce concrete
+    // conformance nor would have any dependent protocols exposed directly,
+    // which means that later assertion is going to be triggered even in
+    // case where type variable is related to archetype which confirms to
+    // required interface.
+    if (auto typeVar = replacement->getAs<TypeVariableType>()) {
```

Comment:
@DougGregor I've changed `solveSingle` to disallow free variables, that worked out pretty well and added comment about that, please take a look.